File: server/routes/telegram.js
// Telegram Bot Integration for Autobook AI
// This module handles integration with the Telegram bot for financial management

export const handleTelegramWebhook = (req, res) => {
  try {
    const { message, update_id } = req.body;

    // Basic webhook validation
    if (!message) {
      return res.status(400).json({
        success: false,
        message: "Invalid webhook data",
      });
    }

    // Process the message (placeholder implementation)
    console.log("Received Telegram message:", message);

    // Here you would typically:
    // 1. Parse the message
    // 2. Extract financial data
    // 3. Process with autobook AI
    // 4. Send response back to user

    res.json({
      success: true,
      message: "Webhook processed successfully",
    });
  } catch (error) {
    console.error("Telegram webhook error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

export const handleSendFinancialUpdate = (req, res) => {
  try {
    const { userId, financialData } = req.body;

    // Validation
    if (!userId || !financialData) {
      return res.status(400).json({
        success: false,
        message: "User ID and financial data are required",
      });
    }

    // Here you would integrate with autobook AI Telegram bot
    // This is a placeholder implementation
    console.log("Sending financial update to Telegram bot:", {
      userId,
      financialData,
    });

    // Simulate sending data to Telegram bot
    const telegramResponse = {
      success: true,
      messageId: Date.now(),
      status: "sent",
    };

    res.json({
      success: true,
      data: telegramResponse,
      message: "Financial update sent to Telegram bot successfully",
    });
  } catch (error) {
    console.error("Send financial update error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

export const handleGetBotStatus = (req, res) => {
  try {
    // Check if bot is connected and operational
    const botStatus = {
      connected: true,
      lastUpdate: new Date().toISOString(),
      activeUsers: 150, // Placeholder
      messagesProcessed: 1247, // Placeholder
      features: [
        "Expense tracking",
        "Invoice processing",
        "Financial alerts",
        "Tax calculations",
        "GST compliance",
      ],
    };

    res.json({
      success: true,
      data: botStatus,
      message: "Bot status retrieved successfully",
    });
  } catch (error) {
    console.error("Get bot status error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

export const handleProcessJobFinancials = (req, res) => {
  try {
    const { jobId, amount, type, description } = req.body;

    // Validation
    if (!jobId || !amount || !type) {
      return res.status(400).json({
        success: false,
        message: "Job ID, amount, and type are required",
      });
    }

    // Process financial data for the job
    const financialData = {
      jobId,
      amount: parseFloat(amount),
      type, // 'income' | 'expense'
      description: description || "",
      processedAt: new Date().toISOString(),
      category:
        type === "income" ? "transportation_revenue" : "operational_expense",
      gstApplicable: amount > 500, // Example logic
      gstAmount: amount > 500 ? amount * 0.18 : 0,
    };

    // Here you would send this data to autobook AI via Telegram
    console.log("Processing job financials:", financialData);

    // Send to Telegram bot for processing
    const telegramMessage = {
      chatId: req.user.telegramChatId || null,
      message: `
🚛 *LogiLedger AI - Financial Update*

💰 Amount: ₹${amount.toLocaleString()}
📊 Type: ${type.toUpperCase()}
📝 Description: ${description || "N/A"}
🏷️ Category: ${financialData.category}
${financialData.gstApplicable ? `\n🧾 GST (18%): ₹${financialData.gstAmount.toFixed(2)}` : ""}

Updated via LogiLedger AI platform.
      `,
    };

    res.json({
      success: true,
      data: financialData,
      telegramMessage,
      message: "Financial data processed and sent to autobook AI",
    });
  } catch (error) {
    console.error("Process job financials error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

export const handleLinkTelegramAccount = (req, res) => {
  try {
    const { telegramUsername, chatId } = req.body;

    if (!telegramUsername) {
      return res.status(400).json({
        success: false,
        message: "Telegram username is required",
      });
    }

    // Store Telegram linking information (in real app, save to database)
    const linkingData = {
      userId: req.user.id,
      telegramUsername,
      chatId: chatId || null,
      linkedAt: new Date().toISOString(),
      status: "pending_verification",
    };

    // Generate verification code
    const verificationCode = Math.random()
      .toString(36)
      .substring(2, 8)
      .toUpperCase();

    res.json({
      success: true,
      data: {
        ...linkingData,
        verificationCode,
      },
      message:
        "Please send this verification code to @autobookAI_bot to complete linking",
    });
  } catch (error) {
    console.error("Link Telegram account error:", error);
    res.status(500).json({
      success: false,
      message: "Internal server error",
    });
  }
};

File: python_backend/routers/telegram.py
from fastapi import APIRouter, HTTPException, status, Depends, Request
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from pydantic import BaseModel
from typing import Optional, Dict, Any
from datetime import datetime
import json
import os
from db import get_users_db, get_jobs_db, get_bids_db, get_consignments_db
from routers.auth import decode_token

router = APIRouter()
security = HTTPBearer()

# Pydantic models
class TelegramWebhook(BaseModel):
    update_id: int
    message: Optional[Dict[str, Any]] = None
    callback_query: Optional[Dict[str, Any]] = None

class FinancialUpdate(BaseModel):
    jobId: str
    financialData: Dict[str, Any]

class BotStatusResponse(BaseModel):
    success: bool
    status: str
    message: str

class LinkAccountRequest(BaseModel):
    telegramUserId: str
    telegramUsername: Optional[str] = None

async def get_current_user(credentials: HTTPAuthorizationCredentials = Depends(security)):
    """Get current user from token"""
    try:
        token = credentials.credentials
        payload = decode_token(token)
        user_id = payload.get("sub")
        
        users_db = get_users_db()
        if user_id not in users_db:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="User not found"
            )
        
        user = users_db[user_id]
        user["id"] = user_id
        return user
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid authentication credentials"
        )

@router.post("/webhook")
async def handle_telegram_webhook(request: Request):
    """Handle Telegram webhook"""
    try:
        # Get the raw body
        body = await request.body()
        data = json.loads(body)
        
        print(f"[TelegramWebhook] Received webhook: {data}")
        
        # Process the webhook based on type
        if "message" in data:
            await process_telegram_message(data["message"])
        elif "callback_query" in data:
            await process_callback_query(data["callback_query"])
        
        return {"status": "ok"}
        
    except Exception as e:
        print(f"[TelegramWebhook] Error processing webhook: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error processing webhook"
        )

async def process_telegram_message(message: Dict[str, Any]):
    """Process incoming Telegram message"""
    try:
        chat_id = message.get("chat", {}).get("id")
        text = message.get("text", "")
        user = message.get("from", {})
        
        print(f"[TelegramMessage] Chat ID: {chat_id}, Text: {text}, User: {user}")
        
        # Handle different commands
        if text.startswith("/start"):
            await send_welcome_message(chat_id)
        elif text.startswith("/help"):
            await send_help_message(chat_id)
        elif text.startswith("/status"):
            await send_status_message(chat_id, user)
        elif text.startswith("/jobs"):
            await send_jobs_message(chat_id, user)
        else:
            await send_default_response(chat_id)
            
    except Exception as e:
        print(f"[TelegramMessage] Error processing message: {e}")

async def process_callback_query(callback_query: Dict[str, Any]):
    """Process callback query from inline keyboards"""
    try:
        chat_id = callback_query.get("message", {}).get("chat", {}).get("id")
        data = callback_query.get("data", "")
        user = callback_query.get("from", {})
        
        print(f"[TelegramCallback] Chat ID: {chat_id}, Data: {data}, User: {user}")
        
        # Handle different callback data
        if data.startswith("job_"):
            job_id = data.split("_")[1]
            await send_job_details(chat_id, job_id, user)
        elif data.startswith("bid_"):
            bid_id = data.split("_")[1]
            await send_bid_details(chat_id, bid_id, user)
        else:
            await send_default_response(chat_id)
            
    except Exception as e:
        print(f"[TelegramCallback] Error processing callback: {e}")

async def send_welcome_message(chat_id: int):
    """Send welcome message to user"""
    message = """
🚛 Welcome to LogiLedger AI Bot!

I'm here to help you manage your logistics operations.

Available commands:
/start - Show this welcome message
/help - Show help information
/status - Check your account status
/jobs - View your active jobs

For support, contact our team.
    """
    # In a real implementation, you would send this via Telegram Bot API
    print(f"[TelegramBot] Sending welcome message to {chat_id}: {message}")

async def send_help_message(chat_id: int):
    """Send help message to user"""
    message = """
📋 LogiLedger AI Bot Help

Commands:
• /start - Welcome message
• /help - This help message
• /status - Your account status
• /jobs - Active jobs

Features:
• Real-time job updates
• Bid notifications
• Financial tracking
• Invoice management

Need help? Contact support.
    """
    print(f"[TelegramBot] Sending help message to {chat_id}: {message}")

async def send_status_message(chat_id: int, user: Dict[str, Any]):
    """Send status message to user"""
    # Find user in database by Telegram ID
    users_db = get_users_db()
    user_record = None
    
    for user_id, user_data in users_db.items():
        if user_data.get("telegramUserId") == str(user.get("id")):
            user_record = user_data
            break
    
    if user_record:
        message = f"""
📊 Account Status

Name: {user_record['name']}
Type: {user_record['userType'].upper()}
Company: {user_record.get('companyName', 'N/A')}
Location: {user_record.get('location', 'N/A')}
        """
    else:
        message = "❌ Account not linked. Please link your account first."
    
    print(f"[TelegramBot] Sending status message to {chat_id}: {message}")

async def send_jobs_message(chat_id: int, user: Dict[str, Any]):
    """Send jobs message to user"""
    # Find user in database
    users_db = get_users_db()
    user_record = None
    
    for user_id, user_data in users_db.items():
        if user_data.get("telegramUserId") == str(user.get("id")):
            user_record = user_data
            break
    
    if not user_record:
        message = "❌ Account not linked. Please link your account first."
    else:
        jobs_db = get_jobs_db()
        
        if user_record["userType"] == "msme":
            # Get MSME jobs
            user_jobs = [
                job for job in jobs_db.values()
                if job["transporterId"] == user_record["id"]
            ]
        else:
            # Get company jobs
            user_jobs = [
                job for job in jobs_db.values()
                if job["companyId"] == user_record["id"]
            ]
        
        if user_jobs:
            message = f"📋 You have {len(user_jobs)} active job(s):\n\n"
            for i, job in enumerate(user_jobs[:5], 1):  # Show first 5 jobs
                message += f"{i}. {job['consignmentTitle']} - {job['status']}\n"
        else:
            message = "📭 No active jobs found."
    
    print(f"[TelegramBot] Sending jobs message to {chat_id}: {message}")

async def send_job_details(chat_id: int, job_id: str, user: Dict[str, Any]):
    """Send job details to user"""
    jobs_db = get_jobs_db()
    
    if job_id in jobs_db:
        job = jobs_db[job_id]
        message = f"""
📋 Job Details

Title: {job['consignmentTitle']}
Status: {job['status']}
Amount: ₹{job['amount']:,.2f}
Origin: {job['origin']}
Destination: {job['destination']}
Deadline: {job['deadline'][:10]}
        """
    else:
        message = "❌ Job not found."
    
    print(f"[TelegramBot] Sending job details to {chat_id}: {message}")

async def send_bid_details(chat_id: int, bid_id: str, user: Dict[str, Any]):
    """Send bid details to user"""
    bids_db = get_bids_db()
    
    if bid_id in bids_db:
        bid = bids_db[bid_id]
        message = f"""
💰 Bid Details

Consignment: {bid['consignmentTitle']}
Amount: ₹{bid['bidAmount']:,.2f}
Status: {bid['status']}
Estimated Delivery: {bid['estimatedDelivery'][:10]}
        """
    else:
        message = "❌ Bid not found."
    
    print(f"[TelegramBot] Sending bid details to {chat_id}: {message}")

async def send_default_response(chat_id: int):
    """Send default response for unrecognized messages"""
    message = "🤖 I didn't understand that. Type /help for available commands."
    print(f"[TelegramBot] Sending default response to {chat_id}: {message}")

@router.post("/financial-update")
async def send_financial_update(
    financial_data: FinancialUpdate,
    current_user: dict = Depends(get_current_user)
):
    """Send financial update to Telegram"""
    try:
        # In a real implementation, you would send this via Telegram Bot API
        print(f"[FinancialUpdate] Sending financial update for job {financial_data.jobId}")
        print(f"[FinancialUpdate] Data: {financial_data.financialData}")
        
        return {
            "success": True,
            "message": "Financial update sent successfully"
        }
        
    except Exception as e:
        print(f"[FinancialUpdate] Error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error sending financial update"
        )

@router.get("/bot-status", response_model=BotStatusResponse)
async def get_bot_status(current_user: dict = Depends(get_current_user)):
    """Get Telegram bot status"""
    try:
        # In a real implementation, you would check the bot's actual status
        return BotStatusResponse(
            success=True,
            status="online",
            message="Telegram bot is running"
        )
        
    except Exception as e:
        return BotStatusResponse(
            success=False,
            status="error",
            message=f"Error checking bot status: {str(e)}"
        )

@router.post("/process-financials")
async def process_job_financials(
    job_id: str,
    current_user: dict = Depends(get_current_user)
):
    """Process job financials and send to Telegram"""
    try:
        jobs_db = get_jobs_db()
        
        if job_id not in jobs_db:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Job not found"
            )
        
        job = jobs_db[job_id]
        
        # Check if user owns this job
        if job["transporterId"] != current_user["id"]:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Access denied"
            )
        
        # Process financials (placeholder)
        financial_data = {
            "jobId": job_id,
            "amount": job["amount"],
            "status": job["status"],
            "processedAt": datetime.now().isoformat()
        }
        
        print(f"[ProcessFinancials] Processed financials for job {job_id}: {financial_data}")
        
        return {
            "success": True,
            "financialData": financial_data,
            "message": "Financials processed successfully"
        }
        
    except Exception as e:
        print(f"[ProcessFinancials] Error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error processing financials"
        )

@router.post("/link-account")
async def link_telegram_account(
    link_request: LinkAccountRequest,
    current_user: dict = Depends(get_current_user)
):
    """Link Telegram account to user"""
    try:
        users_db = get_users_db()
        
        # Update user with Telegram information
        if current_user["id"] in users_db:
            users_db[current_user["id"]]["telegramUserId"] = link_request.telegramUserId
            users_db[current_user["id"]]["telegramUsername"] = link_request.telegramUsername
            users_db[current_user["id"]]["telegramLinkedAt"] = datetime.now().isoformat()
            
            print(f"[LinkAccount] Linked Telegram account {link_request.telegramUserId} to user {current_user['id']}")
            
            return {
                "success": True,
                "message": "Telegram account linked successfully"
            }
        else:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found"
            )
        
    except Exception as e:
        print(f"[LinkAccount] Error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error linking Telegram account"
        ) 